fix: make r/wshada work correct w/ shadafile=NONE

IpsumCapra opened the pull request:
#### Summary
fix: make r/wshada work correct w/ shadafile=NONE

#### Problem
Read/write shada function logic was skipped entirely if it was detected the shadafile option was set to 'NONE'.

#### Solution 
The filename is now always resolved. When the shadafile option is set to 'NONE' AND no filename was passed, the filename resolves to an empty string, which causes the read/write functions to return. Regardless of whether the option is set to 'NONE', when a filename is explicitly passed, it gets resolved and the read/write logic is accessed.

fixes #32510

justinmk commented:
Thanks. Needs a test.

IpsumCapra commented:
Sure, I'll give it a shot. I should probably first figure out why fread_len throws that error during `Test_no_last_substitute_pat()` though.

zeertzjq commented:
Ah, it seems the test only worked by accident, as `Xviminfo` is not a valid shada file.

zeertzjq commented:
Fixing the test is out of scope for this PR, so you can just add a `throw 'skipped: TODO: '` at the top of the test for now:
```diff
diff --git a/test/old/testdir/test_search.vim b/test/old/testdir/test_search.vim
index b62c2b2e5f..f50da79b33 100644
--- a/test/old/testdir/test_search.vim
+++ b/test/old/testdir/test_search.vim
@@ -1449,6 +1449,7 @@ endfunc
 " Test that there is no crash when there is a last search pattern but no last
 " substitute pattern.
 func Test_no_last_substitute_pat()
+  throw 'skipped: TODO: '
   " Use viminfo to set the last search pattern to a string and make the last
   " substitute pattern the most recent used and make it empty (NULL).
   call writefile(['~MSle0/bar', '~MSle0~&'], 'Xviminfo', 'D')
```

justinmk commented on src/nvim/shada.c:
```diff
@@ -1298,7 +1289,11 @@ static char *shada_filename(const char *file)
   if (file == NULL || *file == NUL) {
     if (p_shadafile != NULL && *p_shadafile != NUL) {
-      file = p_shadafile;
+      if (!strequal(p_shadafile, "NONE")) {
```
the old comment had useful information (calling out `--clean`).
```suggestion
      // Check if writing to shada file was disabled ("-i NONE" or "--clean").
      if (!strequal(p_shadafile, "NONE")) {
```

justinmk commented on test/functional/shada/shada_spec.lua:
```diff
@@ -250,6 +250,21 @@ describe('ShaDa support code', function()
   end)
 
+  it(
+    'allows writing and reading ShaDa files when shadafile=NONE but supplied with a filename',
```
```suggestion
    '":wshada/:rshada [filename]" works when shadafile=NONE',
```

justinmk commented on test/functional/shada/shada_spec.lua:
```diff
@@ -250,6 +250,21 @@ describe('ShaDa support code', function()
+    'allows writing and reading ShaDa files when shadafile=NONE but supplied with a filename',
+    function()
+      nvim_command('set shadafile=NONE')
+      eq(0, exc_exec('wshada ' .. shada_fname))
```
`exc_exec` is deprecated, and not needed to check success.
```suggestion
      nvim_command('wshada ' .. shada_fname)
```

justinmk commented on test/functional/shada/shada_spec.lua:
```diff
@@ -250,6 +250,21 @@ describe('ShaDa support code', function()
+      eq(1, read_shada_file(shada_fname)[1].type)
+
+      wshada('Some text file')
+      eq(
+        'Vim(rshada):E576: Error while reading ShaDa file: last entry specified that it occupies 109 bytes, but file ended earlier',
+        exc_exec('rshada ' .. shada_fname)
```
untested, sample code:
```suggestion
      eq(
        'Vim(rshada):E576: Error while reading ShaDa file: last entry specified that it occupies 109 bytes, but file ended earlier',
        t.pcall_err(n.command, 'rshada ' .. shada_fname)
```

IpsumCapra committed:
fix: make r/wshada work correct w/ shadafile=NONE

Problem: read/write shada function logic was skipped entirely if it was
detected the shadafile option was set to 'NONE'.

Solution: The filename is now always resolved. When the shadafile option
is set to 'NONE' AND no filename was passed, the filename resolves to an
empty string, which causes the read/write functions to return.
Regardless of whether the option is set to 'NONE', when a filename is
explicitly passed, it gets resolved and the read/write logic is
accessed.

twhite96 commented:
Getting a shada error after upgrading today. Searched all the closed issues but I haven't seen anything that would help the situation. Pretty certain it was caused by the upgrade as I was just working in nvim a couple minutes before I upgraded. Upgraded and something got corrupted. 

The error: `Error while reading ShaDa file: there is an item at position 24397 that must not be there: Missing items are for internal uses only`

justinmk commented:
How is that related to this PR? Does reverting this commit fix the issue?
